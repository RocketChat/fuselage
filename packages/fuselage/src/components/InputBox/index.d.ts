import { ComponentProps, ForwardRefExoticComponent, ReactNode } from 'react';

import { Box } from '../Box';

type InputBoxProps = ComponentProps<typeof Box> & {
  addon?: ReactNode;
  input?: ReactNode;
  error?: string;
  type:
    | 'button'
    | 'checkbox'
    | 'color'
    | 'date'
    | 'datetime'
    | 'datetime-local'
    | 'email'
    | 'file'
    | 'hidden'
    | 'image'
    | 'month'
    | 'number'
    | 'password'
    | 'radio'
    | 'range'
    | 'reset'
    | 'search'
    | 'submit'
    | 'tel'
    | 'text'
    | 'time'
    | 'url'
    | 'week'
    | 'textarea'
    | 'select';
};
type InputBoxSkeletonProps = ComponentProps<typeof Box>;

export const InputBox: ForwardRefExoticComponent<InputBoxProps> & {
  Skeleton: ForwardRefExoticComponent<InputBoxSkeletonProps>;
};
