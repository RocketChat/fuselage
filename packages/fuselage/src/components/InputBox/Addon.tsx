import type { ComponentProps, Ref } from 'react';
import React, { forwardRef } from 'react';

import Box from '../Box';

type AddonProps = ComponentProps<typeof Box>;

export const Addon = forwardRef(function Addon(
  props: AddonProps,
  ref: Ref<HTMLSpanElement>
) {
  return <Box is='span' rcx-input-box__addon ref={ref} {...props} />;
});
