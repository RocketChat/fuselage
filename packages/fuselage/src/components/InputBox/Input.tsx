import type { ComponentProps, Ref } from 'react';
import React, { forwardRef } from 'react';

import Box from '../Box';

type InputProps = ComponentProps<typeof Box>;

export const Input = forwardRef(function Input(
  props: InputProps,
  ref: Ref<HTMLInputElement>
) {
  return <Box is='input' animated rcx-input-box ref={ref} {...props} />;
});
