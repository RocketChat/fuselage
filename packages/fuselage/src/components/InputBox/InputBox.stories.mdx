import { action } from '@storybook/addon-actions';
import { Meta, Canvas, ArgsTable, Story } from '@storybook/addon-docs/blocks';
import LinkTo from '@storybook/addon-links/react';

import { Icon, InputBox } from '../..';

<Meta title='Forms/InputBox' parameters={{ jest: ['InputBox/spec'] }} />

# InputBox

A decorated input control with support for addons.

Usually you'll perfer to use `-Input` (e.g. <LinkTo kind='Forms/TextInput' story='Default'>`TextInput`</LinkTo>)
components over this one because it works as a construction block for them.

<Canvas>
  <Story name='Default'>
    <InputBox defaultValue='Value' onChange={action('change')} />
  </Story>
</Canvas>

<ArgsTable of={InputBox} />

### With Addon

<Canvas>
  <Story name='With Icon Addon'>
    <InputBox addon={<Icon name='send' size='x20' />} />
  </Story>
</Canvas>

### Invalid

<Canvas>
  <Story name='Invalid'>
    <InputBox defaultValue='Value' error='Error' onChange={action('change')} />
  </Story>
</Canvas>

### Disabled

<Canvas>
  <Story name='Disabled'>
    <InputBox defaultValue='Value' disabled onChange={action('change')} />
  </Story>
</Canvas>

### With placeholder

<Canvas>
  <Story name='With Placeholder'>
    <InputBox placeholder='Placeholder' onChange={action('change')} />
  </Story>
</Canvas>

## InputBox.Skeleton

A skeleton is a placeholder while you wait for content to load, or to visualise content that doesn't exist yet.

<Canvas>
  <Story name='Skeleton'>
    <InputBox.Skeleton />
  </Story>
</Canvas>

<ArgsTable of={InputBox.Skeleton} />
