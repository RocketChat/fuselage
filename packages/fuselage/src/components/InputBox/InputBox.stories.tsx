import { action } from '@storybook/addon-actions';
import {
  Title,
  Description,
  Primary,
  Stories,
  ArgsTable,
} from '@storybook/addon-docs';
import type { ComponentStory, ComponentMeta } from '@storybook/react';
import React from 'react';

import { Icon, InputBox } from '../..';

export default {
  title: 'Inputs/InputBox',
  component: InputBox,
  parameters: {
    docs: {
      description: {
        component:
          "A decorated input control with support for addons. <br/> Usually you'll perfer to use `-Input` (e.g. <LinkTo kind='Forms/TextInput' story='Default'>`TextInput`</LinkTo>) components over this one because it works as a construction block for them.",
      },
      page: () => (
        <>
          <Title />
          <Description />
          <Primary />
          <ArgsTable />
          <Stories title={''} />
        </>
      ),
    },
  },
} as ComponentMeta<typeof Icon>;

const Template: ComponentStory<typeof InputBox> = (args) => (
  <InputBox {...args} />
);

export const Default: ComponentStory<typeof InputBox> = Template.bind({});
Default.args = {
  defaultValue: 'Value',
  onChange: action('change'),
};

export const WithAddon: ComponentStory<typeof InputBox> = Template.bind({});
WithAddon.args = {
  addon: <Icon name='send' size='x20' />,
};

export const Invalid: ComponentStory<typeof InputBox> = Template.bind({});
Invalid.args = {
  defaultValue: 'Value',
  error: 'Error',
  onChange: action('change'),
};

export const Disabled: ComponentStory<typeof InputBox> = Template.bind({});
Disabled.args = {
  defaultValue: 'Value',
  disabled: true,
  onChange: action('change'),
};

export const WithPlaceholder: ComponentStory<typeof InputBox> = Template.bind(
  {}
);
WithPlaceholder.args = {
  placeholder: 'Placeholder',
  onChange: action('change'),
};
export const Skeleton: ComponentStory<typeof InputBox> = () => (
  <InputBox.Skeleton />
);
Skeleton.storyName = 'InputBox.Skeleton';
