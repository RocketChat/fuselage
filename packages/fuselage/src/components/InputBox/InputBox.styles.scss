@use '../../styles/lengths.scss';
@use '../../styles/typography.scss';

.rcx-input-box__wrapper,
%rcx-input-box {
  position: relative;

  display: inline-flex;

  flex-flow: row nowrap;

  align-items: flex-start;

  flex-grow: 1;

  min-width: lengths.size(144);

  padding-block: lengths.padding(8);
  padding-inline: lengths.padding(16) - lengths.border-width(default, rem);

  vertical-align: baseline;

  word-break: break-all;

  outline: 0;

  @extend %input;

  &:invalid,
  &.invalid {
    @extend %input--invalid;
  }

  *:disabled &,
  &:disabled,
  &.disabled {
    cursor: not-allowed;
    pointer-events: none;
  }
}

.rcx-input-box__addon {
  @include clickable;

  flex-flow: row nowrap;

  align-items: flex-start;
  flex: 0 0 auto;
}

@mixin with-colors(
  $color,
  $placeholder-color,
  $focus-caret-color,
  $active-caret-color,
  $disabled-color
) {
  color: #{$color};

  // TODO: fix placeholder performance issue
  // &::placeholder {
  //   color: #{ $placeholder-color };
  // }

  &.rcx-input-box--placeholder-visible {
    color: #{$placeholder-color};
  }

  &:focus,
  &.focus {
    caret-color: #{$focus-caret-color};
  }

  &:active,
  &.active {
    caret-color: #{$active-caret-color};
  }

  *:disabled &,
  &:disabled,
  &.disabled {
    color: #{$disabled-color};
  }
}

@mixin with-icon-addon-colors(
  $color,
  $focus-caret-color,
  $focus-icon-color,
  $disabled-color
) {
  & + .rcx-input-box__addon {
    color: #{$color};
  }

  .rcx-input-box__wrapper.focus > & {
    caret-color: #{$focus-caret-color};
  }

  &:focus + .rcx-input-box__addon,
  &.focus + .rcx-input-box__addon,
  .rcx-input-box__wrapper.focus > & + .rcx-input-box__addon {
    color: #{$focus-icon-color};
  }

  .rcx-input-box__wrapper.disabled > & {
    color: #{$disabled-color};
  }

  *:disabled & + .rcx-input-box__addon,
  &:disabled + .rcx-input-box__addon,
  &.disabled + .rcx-input-box__addon,
  .rcx-input-box__wrapper.disabled > & + .rcx-input-box__addon {
    color: #{$disabled-color};
  }
}

.rcx-input-box {
  position: relative;

  display: inline-flex;

  flex: 1 0 auto;

  min-width: lengths.size(128);

  user-select: initial;

  vertical-align: baseline;

  white-space: nowrap;

  word-break: break-all;

  outline: 0;

  background-color: transparent;

  @include typography.use-font-scale(p2);
  @include typography.use-text-ellipsis;

  &--type-textarea {
    @extend %--with-scrollbars;

    overflow: auto;

    resize: none;
    vertical-align: middle;
    white-space: initial;
  }

  &--type-date,
  &--type-time {
    &::-webkit-inner-spin-button,
    &::-webkit-calendar-picker-indicator {
      position: absolute;
      inset: 0;

      width: auto;
      height: auto;

      cursor: pointer;

      color: transparent;
      background: transparent;
    }
    -webkit-text-fill-color: $input-colors-color;
  }

  &--type-select {
    @extend %--with-scrollbars;

    overflow: auto;
    appearance: none;

    &.rcx-input-box--multiple {
      vertical-align: middle;
    }
  }

  &:not(.rcx-input-box--undecorated) {
    min-width: lengths.size(128);
    min-height: lengths.size(40);

    padding: lengths.padding(8)
      (lengths.padding(16) - lengths.border-width(default, rem));

    @extend %input;

    &:invalid,
    &.invalid {
      @extend %input--invalid;
    }
  }

  &__wrapper:has(.rcx-input-box--small) {
    align-items: center;

    min-width: lengths.size(112);
    max-height: lengths.size(28);
    padding: lengths.padding(4) lengths.padding(16);
  }

  &--small {
    @include typography.use-font-scale(c1);

    &:not(.rcx-input-box--undecorated) {
      @extend %input;
      min-width: lengths.size(112);

      min-height: lengths.size(28);
      padding: lengths.padding(4) lengths.padding(16);

      &:invalid,
      &.invalid {
        @extend %input--invalid;
      }
    }
  }

  @include with-colors(
    $color: $input-colors-color,
    $placeholder-color: $input-colors-placeholder-color,
    $focus-caret-color: $input-colors-focus-caret-color,
    $active-caret-color: $input-colors-active-caret-color,
    $disabled-color: $input-colors-disabled-color
  );

  &:invalid,
  &.invalid {
    @include with-colors(
      $color: $input-colors-invalid-color,
      $placeholder-color: $input-colors-invalid-placeholder-color,
      $focus-caret-color: $input-colors-invalid-focus-caret-color,
      $active-caret-color: $input-colors-invalid-active-caret-color,
      $disabled-color: $input-colors-invalid-disabled-color
    );

    @include with-icon-addon-colors(
      $color: $input-colors-invalid-color,
      $focus-caret-color: $input-colors-invalid-focus-caret-color,
      $focus-icon-color: $input-colors-invalid-focus-icon-color,
      $disabled-color: $input-colors-invalid-disabled-color
    );
  }

  *:disabled &,
  &:disabled,
  &.disabled {
    cursor: not-allowed;
  }

  @include with-icon-addon-colors(
    $color: $input-colors-color,
    $focus-caret-color: $input-colors-focus-caret-color,
    $focus-icon-color: $input-colors-focus-icon-color,
    $disabled-color: $input-colors-disabled-color
  );

  *:disabled & + .rcx-input-box__addon,
  &:disabled + .rcx-input-box__addon,
  &.disabled + .rcx-input-box__addon {
    pointer-events: none;
  }

  .rcx-input-box__wrapper > & {
    width: lengths.size(none);
    min-width: lengths.size(none);

    &--small {
      padding: lengths.padding(none);
    }
  }
}

.rcx-input-box__placeholder {
  @include typography.use-text-ellipsis;
  white-space: nowrap;

  @include typography.use-font-scale(p2);
  color: #{$input-colors-placeholder-color};
}

.rcx-input-box__option {
  @include typography.use-text-ellipsis;
  white-space: nowrap;

  @include typography.use-font-scale(p2);
  color: #{$input-colors-color};
}

.rcx-skeleton__input {
  display: inline-flex;

  flex: 1 0 lengths.size(none);

  min-width: lengths.size(128);
  min-height: lengths.size(40);

  padding: (lengths.padding(12) - lengths.border-width(default, rem))
    (lengths.padding(16) - lengths.border-width(default, rem));

  vertical-align: baseline;

  color: #{$input-colors-color};

  border-width: lengths.border-width(default);
  border-color: #{$input-colors-border-color};
  border-radius: theme('input-border-radius', lengths.border-radius(medium));
  background-color: #{$input-colors-background-color};

  @include typography.use-font-scale(p2);
  @include typography.use-text-ellipsis;
}
