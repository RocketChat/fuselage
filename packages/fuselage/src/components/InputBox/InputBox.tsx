/* eslint-disable complexity */
import { useMergedRefs } from '@rocket.chat/fuselage-hooks';
import type {
  ComponentProps,
  ForwardRefExoticComponent,
  ReactNode,
  Ref,
} from 'react';
import React, { forwardRef, useCallback, useLayoutEffect, useRef } from 'react';

import type { InputBoxSkeleton } from '.';
import { Input, Wrapper } from '.';
import type Box from '../Box';
import { Addon } from './Addon';
import type { Option } from './Option';
import type { Placeholder } from './Placeholder';

type InputBoxProps = ComponentProps<typeof Box> & {
  addon?: ReactNode;
  input?: ReactNode;
  multiple?: boolean;
  error?: string;
  placeholder?: string;
  placeholderVisible?: boolean;
  type:
    | 'button'
    | 'checkbox'
    | 'color'
    | 'date'
    | 'datetime'
    | 'datetime-local'
    | 'email'
    | 'file'
    | 'hidden'
    | 'image'
    | 'month'
    | 'number'
    | 'password'
    | 'radio'
    | 'range'
    | 'reset'
    | 'search'
    | 'submit'
    | 'tel'
    | 'text'
    | 'time'
    | 'url'
    | 'week'
    | 'textarea'
    | 'select';
};

export type InputBox = ForwardRefExoticComponent<InputBoxProps> & {
  Input: ForwardRefExoticComponent<ComponentProps<typeof Box>>;
  Skeleton: ForwardRefExoticComponent<ComponentProps<typeof InputBoxSkeleton>>;
  Option: ForwardRefExoticComponent<ComponentProps<typeof Option>>;
  Placeholder: ForwardRefExoticComponent<ComponentProps<typeof Placeholder>>;
};

// eslint-disable-next-line @typescript-eslint/no-redeclare
export const InputBox = forwardRef(function InputBox(
  {
    className,
    addon,
    error,
    hidden,
    invisible,
    multiple,
    placeholderVisible,
    type = 'text',
    onChange,
    ...props
  }: InputBoxProps,
  ref: Ref<any> | null
) {
  const innerRef = useRef<
    HTMLInputElement | HTMLSelectElement | HTMLTextAreaElement
  >(null);
  const mergedRef = useMergedRefs(ref, innerRef);

  useLayoutEffect(() => {
    if (innerRef.current && innerRef.current.setCustomValidity) {
      innerRef.current.setCustomValidity(error || '');
    }
  }, [error]);

  useLayoutEffect(() => {
    if (addon && innerRef.current && innerRef.current.parentElement) {
      innerRef.current.parentElement.classList.toggle(
        'invalid',
        !innerRef.current.checkValidity()
      );
    }
  }, []);

  const handleChange = useCallback(
    (event) => {
      if (addon && innerRef.current && innerRef.current.parentElement) {
        innerRef.current.parentElement.classList.toggle(
          'invalid',
          !innerRef.current.checkValidity()
        );
      }

      onChange?.call(event.currentTarget, event);
    },
    [addon, onChange]
  );

  if (!addon) {
    return (
      <Input
        is={
          (type === 'textarea' && 'textarea') ||
          (type === 'select' && 'select') ||
          'input'
        }
        className={className}
        cols={(type === 'textarea' && 1) || (type === 'select' && 0) || 0}
        hidden={hidden}
        invisible={invisible}
        multiple={multiple}
        ref={mergedRef}
        htmlSize={
          (type === 'textarea' && undefined) || (type === 'select' && 1) || 1
        }
        type={type === 'textarea' || type === 'select' ? undefined : type}
        onChange={handleChange}
        rcx-input-box--multiple={multiple}
        rcx-input-box--placeholder-visible={placeholderVisible}
        rcx-input-box--type={type}
        {...props}
      />
    );
  }

  return (
    <Wrapper
      className={[
        props.disabled && 'disabled',
        ...(Array.isArray(className) ? className : [className]),
      ]}
      hidden={hidden}
      invisible={invisible}
    >
      <Input
        is={
          (type === 'textarea' && 'textarea') ||
          (type === 'select' && 'select') ||
          'input'
        }
        className={className}
        cols={(type === 'textarea' && 1) || (type === 'select' && 0) || 0}
        multiple={multiple}
        ref={mergedRef}
        htmlSize={
          (type === 'textarea' && undefined) || (type === 'select' && 1) || 1
        }
        type={type === 'textarea' || type === 'select' ? undefined : type}
        onChange={handleChange}
        rcx-input-box--multiple={multiple}
        rcx-input-box--placeholder-visible={placeholderVisible}
        rcx-input-box--type={type}
        rcx-input-box--undecorated
        {...props}
      />
      <Addon children={addon} />
    </Wrapper>
  );
}) as unknown as InputBox;
