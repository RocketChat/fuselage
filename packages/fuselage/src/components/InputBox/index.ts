import { Addon } from './Addon';
import { Input } from './Input';
import { InputBox } from './InputBox';
import { InputBoxSkeleton } from './InputBoxSkeleton';
import { Option } from './Option';
import { Placeholder } from './Placeholder';
import { Wrapper } from './Wrapper';

export * from './Input';
export * from './InputBox';
export * from './InputBoxSkeleton';
export * from './Wrapper';

export default Object.assign(InputBox, {
  Input,
  Skeleton: InputBoxSkeleton,
  Wrapper,
  Addon,
  Placeholder,
  Option,
});
