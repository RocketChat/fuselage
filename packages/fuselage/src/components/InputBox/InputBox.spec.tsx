import { composeStories } from '@storybook/testing-react';
import { render } from '@testing-library/react';
import { axe } from 'jest-axe';

import * as stories from './InputBox.stories';

const { Default } = composeStories(stories);

describe('[InputBox Component]', () => {
  it('renders without crashing', () => {
    const tree = render(<Default />);
    expect(tree.baseElement).toMatchSnapshot();
  });

  it('%s should have no a11y violations', async () => {
    const { container } = render(<Default />);

    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });
});
