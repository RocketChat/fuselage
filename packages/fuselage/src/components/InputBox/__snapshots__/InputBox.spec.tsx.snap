// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[InputBox Component] renders without crashing 1`] = `
<body>
  <div>
    <input
      aria-label="Value"
      class="rcx-box rcx-box--full rcx-box--animated rcx-input-box--type-text rcx-input-box"
      size="1"
      type="text"
      value="Value"
    />
  </div>
</body>
`;
